File: src/main/java/org/profit/server/handler/HttpServerHandler.java
package org.profit.server.handler;

import io.netty.buffer.Unpooled;
import io.netty.channel.ChannelFutureListener;
import io.netty.channel.ChannelHandler.Sharable;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.SimpleChannelInboundHandler;
import io.netty.handler.codec.http.*;
import io.netty.util.CharsetUtil;
import org.profit.server.handler.communication.RequestMessage;
import org.profit.server.handler.communication.exception.InvalidActionException;
import org.profit.server.handler.communication.exception.InvalidParamException;
import org.profit.server.handler.action.Action;
import org.profit.app.Actions;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.Map;

import static io.netty.handler.codec.http.HttpHeaders.Names.CONTENT_TYPE;
import static io.netty.handler.codec.http.HttpMethod.GET;
import static io.netty.handler.codec.http.HttpResponseStatus.FORBIDDEN;
import static io.netty.handler.codec.http.HttpVersion.HTTP_1_1;

/**
 * @author TangYing
 */
@Sharable
public class HttpServerHandler extends SimpleChannelInboundHandler<FullHttpRequest> {

    private static final Logger LOG = LoggerFactory.getLogger(HttpServerHandler.class);

    private HttpStaticFileHandler staticFileHandler = new HttpStaticFileHandler();

    private final static String actionSuffix = ".action";

    @Override
    protected void channelRead0(ChannelHandlerContext ctx, FullHttpRequest request) throws Exception {
        if (request.getMethod() != GET) {
            sendError(ctx, FORBIDDEN);
            return;
        }

        QueryStringDecoder decoder = new QueryStringDecoder(request.getUri());
        Map<String, List<String>> parameters = decoder.parameters();
        String path = decoder.path();
        if (!path.endsWith(actionSuffix)) {
            staticFileHandler.handleStaticFile(ctx, request);
            return;
        }

        // Action request
        String actionUri = getActionUri(path);
        String actionName = actionUri.replaceAll(actionSuffix, "");
        if (actionName.contains("_")) {
            try {
                Action action = Actions.INSTANCE.interpretCommand(actionName);
                if (action == null) {
                    sendError(ctx, FORBIDDEN);
                    return;
                }

                // Execute action
                RequestMessage requestMessage = new RequestMessage(parameters);
                action.execute(ctx, requestMessage);
            } catch (InvalidParamException ipe) {
                LOG.error("InvalidParamException. {}", ipe.getMessage());
                sendError(ctx, FORBIDDEN);
            } catch (InvalidActionException iae) {
                LOG.error("InvalidActionException. {}", iae.getMessage());
                sendError(ctx, FORBIDDEN);
            }
            return;
        }

        // Sya forbidden
        sendError(ctx, FORBIDDEN);
    }

    @Override
    public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
        ctx.flush();
    }

    @Override
    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
        cause.printStackTrace();
        ctx.close();
    }

    /**
     * Send error status
     *
     * @param ctx
     * @param status
     */
    private static void sendError(ChannelHandlerContext ctx, HttpResponseStatus status) {
        FullHttpResponse response = new DefaultFullHttpResponse(HTTP_1_1, status, Unpooled.copiedBuffer("Failure: " + status.toString() + "\r\n", CharsetUtil.UTF_8));
        response.headers().set(CONTENT_TYPE, "text/plain; charset=UTF-8");

        // Close the connection as soon as the error message is sent.
        ctx.writeAndFlush(response).addListener(ChannelFutureListener.CLOSE);
    }

    /**
     * Get action URL
     *
     * @param uri
     * @return
     */
    protected String getActionUri(String uri) {
        String baseUri = getBaseUri(uri);
        return uri.substring(uri.indexOf(baseUri) + baseUri.length());
    }

    /**
     * Get base URL
     *
     * @param uri
     * @return
     */
    protected String getBaseUri(String uri) {
        int idx = uri.indexOf("/", 1);
        if (idx == -1) {
            return "/";
        } else {
            return uri.substring(0, idx);
        }
    }
}

File: src/main/java/org/profit/server/handler/HttpServerHandler.kt
package org.profit.server.handler

import io.netty.buffer.Unpooled
import io.netty.channel.ChannelFutureListener
import io.netty.channel.ChannelHandler.Sharable
import io.netty.channel.ChannelHandlerContext
import io.netty.channel.SimpleChannelInboundHandler
import io.netty.handler.codec.http.*
import io.netty.util.CharsetUtil
import org.profit.app.Actions
import org.profit.server.handler.communication.RequestMessage
import org.profit.server.handler.communication.exception.InvalidActionException
import org.profit.server.handler.communication.exception.InvalidParamException
import org.slf4j.LoggerFactory

/**
 * @author TangYing
 */
@Sharable
class HttpServerHandler : SimpleChannelInboundHandler<FullHttpRequest>() {
    private val staticFileHandler = HttpStaticFileHandler()
    @Throws(Exception::class)
    override fun channelRead0(ctx: ChannelHandlerContext, request: FullHttpRequest) {
        if (request.method !== HttpMethod.GET) {
            sendError(ctx, HttpResponseStatus.FORBIDDEN)
            return
        }
        val decoder = QueryStringDecoder(request.uri)
        val parameters = decoder.parameters()
        val path = decoder.path()
        if (!path.endsWith(actionSuffix)) {
            staticFileHandler.handleStaticFile(ctx, request)
            return
        }
        // Action request
        val actionUri = getActionUri(path)
        val actionName = actionUri.replace(actionSuffix.toRegex(), "")
        if (actionName.contains("_")) {
            try {
                val action = Actions.INSTANCE.interpretCommand(actionName)
                if (action == null) {
                    sendError(ctx, HttpResponseStatus.FORBIDDEN)
                    return
                }
                // Execute action
                val requestMessage = RequestMessage(parameters)
                action.execute(ctx, requestMessage)
            } catch (ipe: InvalidParamException) {
                LOG.error("InvalidParamException. {}", ipe.message)
                sendError(ctx, HttpResponseStatus.FORBIDDEN)
            } catch (iae: InvalidActionException) {
                LOG.error("InvalidActionException. {}", iae.message)
                sendError(ctx, HttpResponseStatus.FORBIDDEN)
            }
            return
        }
        // Sya forbidden
        sendError(ctx, HttpResponseStatus.FORBIDDEN)
    }

    @Throws(Exception::class)
    override fun channelReadComplete(ctx: ChannelHandlerContext) {
        ctx.flush()
    }

    @Throws(Exception::class)
    override fun exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable) {
        cause.printStackTrace()
        ctx.close()
    }

    /**
     * Get action URL
     *
     * @param uri
     * @return
     */
    protected fun getActionUri(uri: String): String {
        val baseUri = getBaseUri(uri)
        return uri.substring(uri.indexOf(baseUri) + baseUri.length)
    }

    /**
     * Get base URL
     *
     * @param uri
     * @return
     */
    protected fun getBaseUri(uri: String): String {
        val idx = uri.indexOf("/", 1)
        return if (idx == -1) {
            "/"
        } else {
            uri.substring(0, idx)
        }
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(HttpServerHandler::class.java)
        private const val actionSuffix = ".action"
        /**
         * Send error status
         *
         * @param ctx
         * @param status
         */
        private fun sendError(ctx: ChannelHandlerContext, status: HttpResponseStatus) {
            val response: FullHttpResponse = DefaultFullHttpResponse(HttpVersion.HTTP_1_1, status, Unpooled.copiedBuffer("Failure: $status\r\n", CharsetUtil.UTF_8))
            response.headers()[HttpHeaders.Names.CONTENT_TYPE] = "text/plain; charset=UTF-8"
            // Close the connection as soon as the error message is sent.
            ctx.writeAndFlush(response).addListener(ChannelFutureListener.CLOSE)
        }
    }
}